### PR TITLE
Implement local prefetch in backend

### PR DESCRIPTION
**Pull request by knizhnik**

### CODE REVIEW

File: contrib/zenith/pagestore_smgr.c [hunk 1]
```diff
@@ -873,9 +918,15 @@ zenith_prefetch(SMgrRelation reln, ForkNumber forknum, BlockNumber blocknum)
...
 
-	/* not implemented */
-	elog(SmgrTrace, "[ZENITH_SMGR] prefetch noop");
-	return true;
+	if (n_prefetch_requests < MAX_PREFETCH_REQUESTS)
+	{
+		prefetch_requests[n_prefetch_requests].rnode = reln->smgr_rnode.node;
+		prefetch_requests[n_prefetch_requests].forkNum = forknum;
+		prefetch_requests[n_prefetch_requests].blockNum = blocknum;
+		n_prefetch_requests += 1;
```
Comment (MMeent) [inline contrib/zenith/pagestore_smgr.c:926]: shouldn't this code block be located in the `case RELPERSISTENCE_PERMANENT` branch above?

File: contrib/zenith/pagestore_client.h [hunk 2]
```diff
@@ -46,6 +46,8 @@ typedef struct
 } ZenithMessage;
 
 #define messageTag(m)		(((const ZenithMessage *)(m))->tag)
+//#define prefetch_log(fmt, ...) elog(LOG, fmt,  ## __VA_ARGS__)
+#define prefetch_log(fmt, ...)
```
Comment (MMeent) [inline contrib/zenith/pagestore_client.h:50]: nit: TODO: cleanup

File: contrib/zenith/pagestore_smgr.c [hunk 3]
```diff
@@ -106,6 +108,49 @@ typedef enum
 static SMgrRelation unlogged_build_rel = NULL;
 static UnloggedBuildPhase unlogged_build_phase = UNLOGGED_BUILD_NOT_IN_PROGRESS;
 
+
+/*
+ * Prefetch implementation:
+ * Prefetch is performed locally by each backend.
+ * There can be up to MAX_PREFETCH_REQUESTS registered using smgr_prefetch
+ * before smgr_read. All this requests are appended to primary smgr_read request.
+ * It is assumed that pages will be requested in prefetch order.
```
Comment (MMeent) [inline contrib/zenith/pagestore_smgr.c:117]: Does this mean that you expect each prefetch request to be requested later, or that when you request a prefetched page, the prefetches before that request will not be requested and can be dropped?

File: src/backend/access/heap/heapam.c [hunk 4]
```diff
@@ -396,9 +397,23 @@ heapgetpage(TableScanDesc sscan, BlockNumber page)
...
+	{
+		int prefetch_limit = seqscan_prefetch_buffers;
+		ParallelBlockTableScanWorker pbscanwork = scan->rs_parallelworkerdata;
+		if (pbscanwork != NULL && pbscanwork->phsw_chunk_remaining < prefetch_limit)
+			prefetch_limit = pbscanwork->phsw_chunk_remaining;
+		if (page + prefetch_limit >= scan->rs_nblocks)
+			prefetch_limit = scan->rs_nblocks - page - 1;
+		for (int i = 1; i <= prefetch_limit; i++)
+			PrefetchBuffer(scan->rs_base.rs_rd, MAIN_FORKNUM, page+i);
+	}
```
Comment (MMeent) [inline src/backend/access/heap/heapam.c:411]: Does the current heap infrastructure not prefetch their buffers?

Additionally, this should probably be gated behind `#ifdef USE_PREFETCH`.

File: src/include/optimizer/cost.h [hunk 5]
```diff
@@ -67,6 +67,9 @@ extern PGDLLIMPORT bool enable_parallel_append;
 extern PGDLLIMPORT bool enable_parallel_hash;
 extern PGDLLIMPORT bool enable_partition_pruning;
 extern PGDLLIMPORT bool enable_async_append;
+extern PGDLLIMPORT bool enable_seqscan_prefetch;
+extern PGDLLIMPORT bool enable_indexscan_prefetch;
```
Comment (MMeent) [inline src/include/optimizer/cost.h:71]: I can't find the usage of this GUC, is it being used?

**Review by MMeent — COMMENTED** [review]
Looks fine, modulo some questions regarding comments and location of the code.

**Comment by knizhnik** [comment]
Results of `pgbench -i -s 1000`:

|prefetch size  |  pgbench -i -s 1000 |
| ------------- | ------------------- |
|             0 | 559.67 s (vacuum 194.02 s, primary keys 168.04 s) |
|             2 | 534.72 s (vacuum 172.22 s, primary keys 142.11 s) |
|             4 | 478.20 s (vacuum 139.28 s, primary keys 130.94 s) |
|             8 | 464.54 s (vacuum 147.98 s, primary keys 125.45 s) |

**Comment by hlinnaka** [comment]
> Results of `pgbench -i -s 1000`:

Thanks! The results are pretty disappointing though :-(. I expected something like a 2-5x speedup. What do you think?

**Comment by knizhnik** [comment]
Yes, I have to agree with you. I also expected about 2 times. But actual improvement is much less.
May be this difference will be larger on real servers (when compute and pageserver are located at different nodes).
Results above were obtained at my laptop where pageserver and compute have to compete for resources. And communication overhead is much smaller in case of local run.

**Comment by knizhnik** [comment]
More results.
Select from `pgbench_accounts` table:
| max_parallel_workers_per_gather | seqscan_prefetch_enabled | selectsum(abaance) from pgbench_accounts |
| --------------------------------------------  | ------------------------------------  | ------------------------------------------------------------ |
|                                                     0 |                                        off  |     147 sec |
|                                                     0 |                                        on  |     91 sec |
|                                                     8 |                                        off  |     56 sec |
|                                                     8 |                                        on  |     53 sec |

CPU usage (no prefetch):
| process | CPU usage |
| pageserver | 59 |
| client backend | 42 |   
| wal redo | 18 |

CPU usage (prefetch):
| process | CPU usage |
| pageserver | 76 |
| client backend | 52 |   
| wal redo | 33 |


Profile of page server:
```
   3,81%  serving Page Se  libc.so.6          [.] __memmove_avx_unaligned_erms
   3,81%  serving Page Se  [kernel.kallsyms]  [k] copy_user_enhanced_fast_string
   2,61%  serving Page Se  pageserver         [.] pageserver::layered_repository::layer_map::LayerMap::search
   2,55%  serving Page Se  pageserver         [.] core::ops::range::Range<Idx>::contains
   2,18%  serving Page Se  [kernel.kallsyms]  [k] psi_group_change
   2,04%  serving Page Se  [kernel.kallsyms]  [k] syscall_exit_to_user_mode
   1,87%  serving Page Se  pageserver         [.] <pageserver::layered_repository::image_layer::ImageLayer as pageserver::layered_repository::storage_layer
   1,48%  serving Page Se  libc.so.6          [.] _int_malloc
   1,44%  serving Page Se  libc.so.6          [.] pthread_rwlock_rdlock@GLIBC_2.2.5
   1,42%  serving Page Se  pageserver         [.] pageserver::page_cache::PageCache::try_lock_for_read
   1,32%  serving Page Se  pageserver         [.] <pageserver::layered_repository::delta_layer::DeltaLayer as pageserver::layered_repository::storage_layer
   1,09%  serving Page Se  libc.so.6          [.] _int_free
   1,09%  serving Page Se  libc.so.6          [.] pthread_rwlock_unlock@@GLIBC_2.34
   1,00%  serving Page Se  [kernel.kallsyms]  [k] syscall_return_via_sysret
```

So looks like prefetch doesn't give signficant improvement because bottleneck is pagerserver: it is just not able to fetch/reconstruct pages fast enough (in one stream). Certainly we can add parallel prefetch support on pageserver side (execute prefetch requests in separate thread). But I am not sure how good this solution will work in multitenent environment (it will increase risk that one tenant consumes all system resources, slowing down other tenants).

And  any case parallel execution can reduce speed only two times.

Also please notice that at this volume (scale 1000: ~10Gb), sequential lookup in layer map becomes expensive.
One again want to notice PR with using R-Tree for it.

I need to notice that `pgbench -i` is using copy which is storing FPI whih should be stored as images in pageserver store. But still we see that a lot of time is consumed in wal-red. And there is almost no way now to force pageserver to generate page images.

**Comment by knizhnik** [comment]
I tried to eliinate all page reconstruction overhead from page server (just patch ` get_rel_page_at_lsn` to always request first page of `pgbench_accounts` relation, which is cached). The query execution time is reduced on time of wal-redo, but common picture is not changed:

max_parallel_workers_per_gather | seqscan_prefetch_enabled | selectsum(abaance) from pgbench_accounts
-- | -- | --
0 | off | 116 sec
0 | on | 56 sec
8 | off | 30 sec
8 | on | 23 sec

So now prefetch provides more than two times improvement for sequential scan.
For parallel plan the improvement is not so large.

Profile of page server:
```
  13,14%  serving Page Se  pageserver         [.] pageserver::layered_repository::layer_map::LayerMap::search
   8,13%  serving Page Se  pageserver         [.] core::ops::range::Range<Idx>::contains
   4,36%  serving Page Se  pageserver         [.] <pageserver::layered_repository::delta_layer::DeltaLayer as pageserver::layered_repository::storage_lay
   3,99%  serving Page Se  libc.so.6          [.] pthread_rwlock_rdlock@GLIBC_2.2.5
   3,66%  serving Page Se  pageserver         [.] <pageserver::layered_repository::image_layer::ImageLayer as pageserver::layered_repository::storage_lay
   3,27%  serving Page Se  libc.so.6          [.] __memmove_avx_unaligned_erms
   2,59%  serving Page Se  pageserver         [.] <pageserver::layered_repository::delta_layer::DeltaLayer as pageserver::layered_repository::storage_lay
   2,47%  serving Page Se  pageserver         [.] <pageserver::layered_repository::image_layer::ImageLayer as pageserver::layered_repository::storage_lay
   2,22%  serving Page Se  pageserver         [.] pageserver::page_cache::PageCache::try_lock_for_read
   2,20%  serving Page Se  libc.so.6          [.] __memcmp_avx2_movbe
   2,04%  serving Page Se  libc.so.6          [.] pthread_rwlock_unlock@@GLIBC_2.34
   1,65%  serving Page Se  pageserver         [.] <pageserver::layered_repository::timeline::LayeredTimeline as pageserver::repository::Timeline>::get
   1,48%  serving Page Se  pageserver         [.] pageserver::layered_repository::disk_btree::DiskBtreeReader<R,_>::search_recurse
   1,25%  serving Page Se  pageserver         [.] <pageserver::layered_repository::delta_layer::DeltaLayer as pageserver::layered_repository::storage_lay
   1,02%  serving Page Se  libc.so.6          [.] _int_free
   1,01%  serving Page Se  libc.so.6          [.] _int_malloc
   0,94%  serving Page Se  pageserver         [.] pageserver::layered_repository::disk_btree::OnDiskNode<_>::binary_search
```

Please notice 21% (13+8) spent in layer map lookup.

**Comment by knizhnik** [comment]
Replaced with #202

**Pull request closed without merging**